android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt

```diff
@@ -0,0 +1,63 @@
+package com.badoo.ribs.android
+
+import com.badoo.ribs.android.RequestCodeBasedEventStream.RequestCodeBasedEvent
+import com.badoo.ribs.core.Identifiable
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import com.jakewharton.rxrelay2.Relay
+import io.reactivex.Observable
+
+abstract class RequestCodeBasedEventStreamImpl<T : RequestCodeBasedEvent>(
+    private val requestCodeRegistry: RequestCodeRegistry
+) : RequestCodeBasedEventStream<T> {
+    private val events = HashMap<Int, Relay<T>>()
+
+    override fun events(client: Identifiable): Observable<T> {
+        val id = requestCodeRegistry.generateGroupId(client.id)
+        ensureSubject(id)
+
+        return events.getValue(id)
+            .doOnDispose { cleanup(id) }
+            .hide()
+    }
+
+    private fun ensureSubject(id: Int, onSubjectDidNotExist: (() -> Unit)? = null) {
```

Comment:
Rename `onSubjectDidNotExist` to `init`?

Reply:
It's used as a point to throw exceptions / log stuff, not to init

Reply:
Yeah, but from this method perspective it does not make sense

Reply:
You can also make it return `Boolean` saying if the subject was created or not.

Reply:
I started with boolean. The current one makes for a more expressive usage for me.

Reply:
+ It's purely internal. Keeping it.